### PR TITLE
Simplify storage interface

### PR DESCRIPTION
**Pull request by wojtek-t**
```release-note
NONE
```

/kind cleanup
/priority backlog
/sig api-machinery

### CODE REVIEW

**Comment by wojtek-t** [comment]
/assign @liggitt 

@liggitt - would you mind taking a look at this? I would like to get your thoughs about the safety of this change and if we need to maybe do something to avoid some potential bugs due to it.

**Comment by wojtek-t** [comment]
/assign @deads2k 

@deads2k - can you share your thoughts about potential impact on some downstream implementations? I would like to have this merged to simplify the code a bit, but I would like to ensure that it's not too misleading...
Alternatively, maybe we should get rid of "List" method, and leave just "GetToList" (where forcing to set recursive seems more natural)?

**Comment by wojtek-t** [comment]
> Alternatively, maybe we should get rid of "List" method, and leave just "GetToList" (where forcing to set recursive seems more natural)?

Or maybe we should just replace the two existing methods (List and GetToList) with a new one GetList (correspondence with Get method), where the fact of new semantic will be pretty obvious.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 1]
```diff
@@ -635,11 +617,13 @@ func (c *Cacher) listItems(listRV uint64, key string, pred storage.SelectionPred
 	return c.watchCache.WaitUntilFreshAndList(listRV, pred.MatcherIndex(), trace)
 }
 
-func (c *Cacher) list(ctx context.Context, key string, opts storage.ListOptions, listObj runtime.Object, recursive bool) error {
+// List implements storage.Interface.
+func (c *Cacher) List(ctx context.Context, key string, opts storage.ListOptions, listObj runtime.Object) error {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:621]: this change means that existing callers of `List(ctx context.Context, key string, opts storage.ListOptions, listObj runtime.Object)` that don't know about the new Recursive field when they construct their `ListOptions` struct will break, right?

File: pkg/controlplane/reconcilers/lease.go [hunk 2]
```diff
@@ -67,6 +67,7 @@ func (s *storageLeases) ListLeases() ([]string, error) {
 		ResourceVersion:      "0",
 		ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan,
 		Predicate:            storage.Everything,
+		Recursive:            true,
```
Comment (liggitt) [inline pkg/controlplane/reconcilers/lease.go:70]: this is a good example of the usage change that is surprising/unfortunate

**Comment by liggitt** [comment]
> Or maybe we should just replace the two existing methods (List and GetToList) with a new one GetList (correspondence with Get method), where the fact of new semantic will be pretty obvious.

If we're going to change the meaning of existing invocations, doing it in a way that makes callers notice seems better than silently breaking existing `List` invocations

File: pkg/controlplane/reconcilers/lease.go (hunk 2 above)
Comment (wojtek-t) [reply pkg/controlplane/reconcilers/lease.go:70]: Renamed the new function to something new, so that every usage has to be explicitly migrated.

**Comment by wojtek-t** [comment]
/retest

Unrelated flake: https://github.com/kubernetes/kubernetes/issues/108043

**Comment by wojtek-t** [comment]
@liggitt @deads2k - PTAL

**Comment by liggitt** [comment]
the current state won't silently break someone, which is good, but it's not super clear to me what the benefit of this change is... making a single function call return two possible things seems harder to understand

**Comment by wojtek-t** [comment]
@deads2k @liggitt - PTAL

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go [hunk 3]
```diff
@@ -183,24 +183,18 @@ type Interface interface {
...
+	// Get unmarshals object found at key into objPtr. On a not found error, will either
 	// return a zero object of the requested type, or an error, depending on 'opts.ignoreNotFound'.
 	// Treats empty responses and nil response nodes exactly like a not found error.
 	// The returned contents may be delayed, but it is guaranteed that they will
 	// match 'opts.ResourceVersion' according 'opts.ResourceVersionMatch'.
 	Get(ctx context.Context, key string, opts GetOptions, objPtr runtime.Object) error
 
-	// GetToList unmarshals json found at key and opaque it into *List api object
-	// (an object that satisfies the runtime.IsList definition).
+	// GetList unmarshalls objects found at key range and opaque them into *List api
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:193]:
```suggestion
	// GetList unmarshalls objects found at key into a *List api
```

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go [hunk 4]
```diff
@@ -183,24 +183,18 @@ type Interface interface {
...
+	// object (an object that satisfies runtime.IsList definition).
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:194]: Add something like "If opts.Recursive is false, 'key' is used as an exact match. If opts.Recursive is true, 'key' is used as a prefix."?

**Review by liggitt — COMMENTED** [review]
a few doc clarifications/suggestions, looks reasonable otherwise

**Comment by wojtek-t** [comment]
@liggitt - thanks for the review, comments applied - PTAL

**Pull request merged**